lib/workload.hpp

```diff
@@ -127,98 +129,146 @@ ostream& operator<<(ostream& os, const rhs_t& rhs);
 bool operator==(const rhs_t& rhs1, const rhs_t& rhs2);
 bool operator<(const rhs_t& rhs1, const rhs_t& rhs2);
 
+//************************************* WlSpec *************************************//
+
+class WlSpec {
+public:
+    virtual ~WlSpec() = default;
+    virtual bool applies_to_queue(unsigned int queue) const = 0;
+    virtual bool equals(const WlSpec& other) const = 0;
+    virtual bool less_than(const WlSpec& other) const = 0;
+    virtual int type_id() const = 0;
+    virtual std::string to_string() const = 0;
+    virtual bool is_available_for_search() const { return false; }
+};
+
+//typedef variant<Comp, Same, Incr, Decr, Unique> wl_spec_t;
+
+bool wl_spec_applies_to_queue(const WlSpec& spec, unsigned int queue);
+bool wl_spec_applies_to_queue(const std::shared_ptr<WlSpec>& spec, unsigned int queue);
+
+bool wl_spec_is_empty(const WlSpec& spec);
+bool wl_spec_is_empty(const std::shared_ptr<WlSpec>& spec);
+
+bool wl_spec_is_all(const WlSpec& spec);
+bool wl_spec_is_all(const std::shared_ptr<WlSpec>& spec);
+
+unsigned int wl_spec_ast_size(const WlSpec& spec);
+unsigned int wl_spec_ast_size(const std::shared_ptr<WlSpec>& wl_spec);
```

Comment:
Why have these methods not moved to WlSpec (single instance ones)?

Reply:
Should they be static methods? Ex:
`static bool wl_spec_applies_to_queue(WlSpec* spec, unsigned int queue)`

Or something like this:
`bool applies_to_queue(unsigned int queue)`

Reply:
Second option. The point of moving them to class is removing the WlSpec parameter

Reply:
[Move global functions and operator overloads to WlSpec class](https://github.com/all-things-networking/fperf/pull/27/commits/252c251ed12899321f4f132b2b40a64deae5b995)